### PR TITLE
Adds SHA256 Cryptocell code to bootloader

### PR DESCRIPTION
**Pull request by kaczmarczyck**
The last piece to make the bootloader actually work. Advantage of a hardware SHA256 is the much faster boot time.

### CODE REVIEW

File: bootloader/src/bitfields.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
+// Copyright 2020-2021 Google LLC
```
Comment (jmichelp) [inline bootloader/src/bitfields.rs]: 2022 now :)

File: bootloader/src/crypto_cell.rs [hunk 2]
```diff
@@ -0,0 +1,283 @@
+// Copyright 2019-2021 Google LLC
```
Comment (jmichelp) [inline bootloader/src/crypto_cell.rs]: Same here

File: bootloader/src/crypto_cell.rs [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    fn clear_data(&self) {
+        let mut ctx = self.hash_ctx.get();
+        ctx.iter_mut().for_each(|b| *b = 0);
+        self.hash_ctx.set(ctx);
+        self.hash_total_size.set(0);
+    }
+
+    /// Adds data to the current hash computation.
+    ///
+    /// You have to know in advance is this is going to be the last block, and indicate that
```
Comment (jmichelp) [inline bootloader/src/crypto_cell.rs]: typo: advance if this is going

File: bootloader/src/registers.rs [hunk 4]
```diff
@@ -0,0 +1,139 @@
+// Copyright 2020-2021 Google LLC
```
Comment (jmichelp) [inline bootloader/src/registers.rs]: Again :)

**Review by jmichelp — COMMENTED** [review]
As we may need the Cryptocell code also in Tock, I wonder if this wouldn't make sense to make it a separate crate that we could import in both the bootloader and the kernel. Otherwise we'll end up with code duplication and different kinds of API to compute hashes.

**Comment by kaczmarczyck** [comment]
Question is if you want the new crate now or submit and put the work in when / if we submit it to Tock?

This PR is simpler than what I think you had in mind for Tock, i.e. it's API doesn't allow concurrent calls. So I don't know if a simply move to a new crate would actually be that useful?

**Comment by jmichelp** [comment]
We can merge it as-is at the moment to get the feature out.
And let's align with @ia0 to make it a separate crate that can bootstrap the work around hardware crypto support.

**Review by jmichelp — APPROVED** [review]

**Comment by ia0** [comment]
> And let's align with @ia0 to make it a separate crate that can bootstrap the work around hardware crypto support.

Ack! (I don't have any ideas yet, and that might be related to ctap as a library too, so should definitely be in our radar).

**Comment by kaczmarczyck** [comment]
Thanks!

**Pull request merged**